indy-besu/smart_contracts/contracts/cl/CredentialDefinitionTypes.sol

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+/**
+ * @title CredentialDefinitionWithMetadata
+ * @dev This struct holds the details of a credential definition
+ * and its associated metadata.
+ *
+ * @param credDef - The details of the credential definition.
+ * @param metadata - Additional metadata associated with the credential definition.
+ */
+struct CredentialDefinitionWithMetadata {
+    CredentialDefinition credDef;
+    CredentialDefinitionMetadata metadata;
+}
+
+/**
+ * @title CredentialDefinition
+ * @dev This struct holds the essential details of a credential definition.
+ *
+ * @param id - Unique identifier for the credential definition.
+ * @param issuerId - Identifier for the issuer of the credential.
+ * @param schemaId - Identifier for the schema associated with this credential definition.
+ * @param credDefType - Type or of the credential definition.
+ * @param tag - A tag or label associated with the credential definition.
+ * @param value - The value of the credential definition.
+ */
+struct CredentialDefinition {
+    string id;
```

Comment:
There are two ways of defining CL-entities:
- DID URL - did:indy:indy2:5nDyJVP1NrcPAttP3xwMB9/anoncreds/v0/CLAIM_DEF/56495/npdb
- Legacy identifier - 5nDyJVP1NrcPAttP3xwMB9:3:CL:56495:npdb
For making migration possible we have to support original Indy identifiers.
In your example `XTbXnccif2tMn3y4bfKCsP:3:CL:65770:acme` the `3` is the enumerated object type that we don't use in Indy2. [More about original Indy identifiers.](https://hyperledger.github.io/indy-did-method/#cred-def)

Reply:
> We decided to require passing id property for schemas and creddef to store exactly the same object (state proof) as will be retuned on resolving.

I suppose mainly this answers the question. I have touched solidity in years, but conceptually I dislike redundancy - it's in:

1. mapping id 
```
mapping(string id => CredentialDefinitionWithMetadata credDefWithMetadata) private _credDefs;
```

2. it's in `id` attribute
```
struct CredentialDefinition {
    string id;
    ...
}
```
3. and broke down in individual piece of the struct as well 
```
struct CredentialDefinition {
    string issuerId;
    string credDefType;
    string tag;
    ....
```

This can be also just matter of taste, but I personally find less redundancy more favorable.

Reply:
In general - yes. But usually entities without obvious identifier, just with a natural key are not a good pattern.